libraries/Microsoft.Bot.Builder.Core.Extensions/ShowTypingMiddleware.cs

```diff
@@ -0,0 +1,55 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Core.Extensions
+{
+    public class ShowTypingMiddleware : IMiddleware
+    {
+        /// <summary>
+        /// (Optional) initial delay before sending first typing indicator. Defaults to 500ms.
+        /// </summary>
+        private readonly int _delay;
+
+        /// <summary>
+        /// (Optional) rate at which additional typing indicators will be sent. Defaults to every 2000ms.
+        /// </summary>
+        private readonly int _freqency;
+
+        public ShowTypingMiddleware(int delay = 500, int frequency = 2000)
+        {
+            _delay = delay;
+            _freqency = frequency;
+        }
+
+        public async Task OnProcessRequest(IBotContext context, MiddlewareSet.NextDelegate next)
```

Comment:
This is awesome, but the corner cases are making my head hurt. 

I *think* everything you have here is right, and the scope of the timer mirrors the scope of the context.